pytorch_lightning/metrics/functional/classification.py

```diff
@@ -872,6 +873,89 @@ def average_precision(
     return -torch.sum((recall[1:] - recall[:-1]) * precision[:-1])
 
 
+def object_detection_mean_average_precision(
+    pred: torch.Tensor, target: torch.Tensor, iou_threshold: float, num_classes: int
+) -> torch.Tensor:
+    """
+    Compute mean average precision for object detection task. Calculates average precision
+    using AUC.
+
+    Args:
+        pred: Tensor containing predictions, where each row is of the following format:
+              [image_idx, class_pred, class_prob, x_min, y_min, x_max, y_max]
+        target: Tensor containing ground truths, where each row is of the follwing format:
+                [image_idx, class_label, x_min, y_min, x_max, y_max]
+        iou_threshold: threshold for IoU score for determining true positive and
+                       false positive predictions.
+        num_classes: number of total classes
+
+    Returns:
+        mean of the average precision for each class in object detection task.
+    """
+    average_precisions = torch.zeros(num_classes)
+    for c in range(num_classes):
+        c_pred = [p for p in pred if p[1] == c]
+        c_target = [t for t in target if t[1] == c]
+        if len(c_target) == 0:
+            continue
+        # Sort c_pred in descending order w.r.t. confidence score
+        c_pred = sorted(c_pred, key=lambda x: x[2], reverse=True)
+        num_targets_per_image = Counter(t[0].item() for t in c_target)
+        targets_tracker = {
+            image_idx: torch.zeros(count)
+            for image_idx, count in num_targets_per_image.items()
+        }
+        tps, fps = torch.zeros(len(c_pred)), torch.zeros(len(c_pred))
+        for i, p in enumerate(c_pred):
+            # Get the targets that correspond to the same image as the prediction
+            ground_truths = [t for t in c_target if t[0] == p[0]]
+            best_iou = 0
+            best_target_idx = 0
+            for j, t in enumerate(ground_truths):
+                curr_iou = object_detection_iou(p[None, 3:], t[None, 2:])
```

Comment:
I added torchvision's `box_iou` instead of using a custom implementation!

Reply:
Great! By the way, `box_iou()` is able to calculate a matrix of IoUs given two sets of boxes. I don't know if it would be faster to calculate a matrix of IoUs between all predicted and target box pairs once, instead of calculating all the pairs individually. At least it's possible that the function is optimized so that it won't compute every pair twice. Then just sort the rows to get the best target for each prediction.  Could be a slightly tidier implementation too. What do you think?

Reply:
You raise a good point! A large amount of the calculation might be unnecessary since it will calculate IoUs between prediction bounding boxes and target bounding boxes of different images and different labels, but the vectorization in itself might lead to better performance despite doing more work than necessary. I do think the way it's implemented right now is probably easier to understand without compromising too much speed. I'll definitely implement your idea and compare. What are you thoughts on that?

Reply:
I have no idea how much speed can be gained, but I know the AP computation can take a lot of time, so if you can try it, it would be great!

Reply:
I just changed the implementation as you suggested!